misc/genstd/util.go

```diff
@@ -0,0 +1,119 @@
+package main
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"os/exec"
+	"path"
+	"path/filepath"
+	"strings"
+	"sync"
+)
+
+func runTool(importPath string) error {
+	shortName := path.Base(importPath)
+	gr := gitRoot()
+
+	cmd := exec.Command(
+		"go", "run", "-modfile", filepath.Join(gr, "misc/devdeps/go.mod"),
+		importPath, "-w", "native.go",
+	)
+	_, err := cmd.Output()
+	if err != nil {
+		if err, ok := err.(*exec.ExitError); ok {
+			return fmt.Errorf("error executing %s: %w; output: %v", shortName, err, string(err.Stderr))
+		}
+		return fmt.Errorf("error executing %s: %w", shortName, err)
+	}
+	return nil
+}
+
+var (
+	memoGitRoot string
+	memoRelPath string
+
+	dirsOnce sync.Once
+)
+
+func gitRoot() string {
+	dirsOnceDo()
+	return memoGitRoot
+}
+
+func relPath() string {
+	dirsOnceDo()
+	return memoRelPath
+}
+
+func dirsOnceDo() {
+	dirsOnce.Do(func() {
+		var err error
+		memoGitRoot, memoRelPath, err = findDirs()
+		if err != nil {
+			panic(fmt.Errorf("could not determine git root: %w", err))
+		}
+	})
+}
+
+func findDirs() (gitRoot string, relPath string, err error) {
+	wd, err := os.Getwd()
+	if err != nil {
+		return
+	}
+	p := wd
+	for {
+		if s, e := os.Stat(filepath.Join(p, ".git")); e == nil && s.IsDir() {
+			// make relPath relative to the git root
+			rp := strings.TrimPrefix(wd, p+string(filepath.Separator))
+			// normalize separator to /
+			rp = strings.ReplaceAll(rp, string(filepath.Separator), "/")
+			return p, rp, nil
+		}
+
+		if strings.HasSuffix(p, string(filepath.Separator)) {
+			return "", "", errors.New("root git not found")
+		}
+
+		p = filepath.Dir(p)
+	}
+}
+
+// pkgNameFromPath derives the package name from the given path,
+// unambiguously for the most part (so safe for the code generation).
+//
+// The path is taken and possibly shortened if it starts with a known prefix.
+// For instance, github.com/gnolang/gno/stdlibs/std simply becomes "libs_std".
+// "Unsafe" characters are removed (ie. invalid for go identifiers).
+func pkgNameFromPath(path string) string {
```

Comment:
Let's put this logic inside gnovm/pkg/gnolang/misc.go, and replacing the impl of `gnovm/pkg/gnolang/misc.go:func defaultPkgName(gopkgPath string) Name {`, but keeping the return type as Name, and also replacing  `gnovm/tests/file.go:func DefaultPkgName(gopkgPath string) gno.Name {`, so instead of 3 we have 1 impl.

Reply:
I don't know, I'm not convinced by this. They address different things: this one is for shortening and "canonicizing" Go import paths, while defaultPkgName operates on Gno import paths. Furthermore, I tried doing what you said but now I start having breaking tests and expect this is not the only one.

![image](https://github.com/gnolang/gno/assets/4681308/6ff5d7fd-2fa8-4670-9e24-acc0474dc9cb)

Reply:
ok